app/src/main/java/me/timpushkin/sgbmandroidapp/MainActivity.kt

```diff
@@ -0,0 +1,155 @@
+package me.timpushkin.sgbmandroidapp
+
+import android.graphics.Bitmap
+import android.graphics.BitmapFactory
+import android.net.Uri
+import android.os.Bundle
+import android.os.SystemClock
+import androidx.activity.ComponentActivity
+import androidx.activity.compose.setContent
+import androidx.compose.foundation.Image
+import androidx.compose.foundation.layout.Arrangement
+import androidx.compose.foundation.layout.Column
+import androidx.compose.foundation.layout.fillMaxSize
+import androidx.compose.foundation.layout.padding
+import androidx.compose.material.MaterialTheme
+import androidx.compose.material.Scaffold
+import androidx.compose.material.Text
+import androidx.compose.material.rememberScaffoldState
+import androidx.compose.runtime.*
+import androidx.compose.runtime.saveable.rememberSaveable
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.graphics.asImageBitmap
+import kotlinx.coroutines.launch
+import me.timpushkin.sgbmandroid.DepthEstimator
+import me.timpushkin.sgbmandroidapp.ui.MenuButtons
+import me.timpushkin.sgbmandroidapp.ui.ScaleFactorField
+import me.timpushkin.sgbmandroidapp.utils.StorageUtils
+import me.timpushkin.sgbmandroidapp.utils.depthArrayToBitmap
+
+private const val CACHED_PARAMS = "params.xml"
+private const val CACHED_LEFT_IMAGE = "left.png"
+private const val CACHED_RIGHT_IMAGE = "right.png"
+private const val CACHED_DEPTH_MAP = "depth.png"
+
+class MainActivity : ComponentActivity() {
+    private lateinit var mStorageUtils: StorageUtils
+    private lateinit var mDepthEstimator: DepthEstimator
+
+    private var mImages: Pair<ByteArray, ByteArray>? = null
+    private var mDepthMap by mutableStateOf<Bitmap?>(null)
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        mStorageUtils = StorageUtils(this)
+        with(mStorageUtils) {
+            getFromCache(CACHED_PARAMS)?.run { mDepthEstimator = DepthEstimator(path) }
+            getFromCache(CACHED_LEFT_IMAGE)?.let { left ->
+                getFromCache(CACHED_RIGHT_IMAGE)?.let { right ->
+                    mImages = left.readBytes() to right.readBytes()
+                }
+            }
+            getFromCache(CACHED_DEPTH_MAP)?.run { mDepthMap = BitmapFactory.decodeFile(path) }
+        }
+
+        setContent {
+            MaterialTheme {
```

Comment:
It looks not really great that all logic and ui definition is here, level of nesting is also very very high

Reply:
I can agree with this, but I tried to do it as much UI-only as possible, or it seems to me like this at least.

I rewrote the app, so now it is a little bit less complicated. I can also extract the whole code inside `MaterialTheme` into a separate composable (something like `MainScreen`), but then I'll have to pass some `MainAcitvity` functions to it.